File: Sources/PalmierPro/Timeline/SnapEngine.swift
import Foundation

enum SnapEngine {

    // MARK: - Types

    struct SnapTarget {
        let frame: Int
        let kind: Kind
        enum Kind { case playhead, clipEdge }
    }

    struct SnapResult {
        let frame: Int
        let probeOffset: Int // which probe snapped (0=start, duration=end)
        let x: Double // snap indicator pixel position
    }

    /// Mutable state that persists across drag events for sticky snap behavior.
    struct SnapState {
        var currentlySnappedTo: Int?
        var currentProbeOffset: Int = 0 // which probe is sticky
    }

    // MARK: - Target collection

    /// Collects all clip edges and the playhead as snap targets.
    /// Optionally excludes specific clip IDs (e.g., the clip being dragged).
    static func collectTargets(
        tracks: [Track],
        playheadFrame: Int,
        excludeClipIds: Set<String> = []
    ) -> [SnapTarget] {
        var targets = [SnapTarget(frame: playheadFrame, kind: .playhead)]
        for track in tracks {
            for clip in track.clips where !excludeClipIds.contains(clip.id) {
                targets.append(SnapTarget(frame: clip.startFrame, kind: .clipEdge))
                targets.append(SnapTarget(frame: clip.endFrame, kind: .clipEdge))
            }
        }
        return targets
    }

    // MARK: - Snap finding

    /// Snap position(s) to nearest target, with sticky behavior and playhead priority.
    /// Tests one or more probe positions (e.g., clip start and end) against all targets.
    static func findSnap(
        position: Int,
        probeOffsets: [Int] = [0],
        targets: [SnapTarget],
        state: inout SnapState,
        baseThreshold: Double,
        pixelsPerFrame: Double
    ) -> SnapResult? {
        let baseFrameThreshold = baseThreshold / pixelsPerFrame

        // Sticky: stay snapped until moved 2.5x threshold away
        if let snapped = state.currentlySnappedTo {
            let holdThreshold = baseFrameThreshold * Snap.stickyMultiplier
            let probePos = position + state.currentProbeOffset
            if abs(Double(probePos - snapped)) <= holdThreshold,
               targets.contains(where: { $0.frame == snapped }) {
                return SnapResult(frame: snapped, probeOffset: state.currentProbeOffset, x: Double(snapped) * pixelsPerFrame)
            }
            state.currentlySnappedTo = nil
            state.currentProbeOffset = 0
        }

        // Find closest (probe, target) pair
        var best: (probeOffset: Int, target: SnapTarget, distance: Double)?
        for probeOffset in probeOffsets {
            let probePos = position + probeOffset
            for target in targets {
                let threshold: Double = switch target.kind {
                case .playhead: baseFrameThreshold * Snap.playheadMultiplier
                case .clipEdge: baseFrameThreshold
                }
                let dist = abs(Double(probePos - target.frame))
                if dist <= threshold, dist < (best?.distance ?? .infinity) {
                    best = (probeOffset, target, dist)
                }
            }
        }

        guard let best else { return nil }
        state.currentlySnappedTo = best.target.frame
        state.currentProbeOffset = best.probeOffset
        return SnapResult(frame: best.target.frame, probeOffset: best.probeOffset, x: Double(best.target.frame) * pixelsPerFrame)
    }

}

File: platforms/windows/src/PalmierPro.App/Editing/SnapEngine.cs
using PalmierPro.Core.Models;

namespace PalmierPro.App.Editing;

/// Pure timeline-drag snap math, ported near-verbatim from Timeline/SnapEngine.swift. No WinUI
/// dependency — lives in `PalmierPro.App` (not `PalmierPro.Core`) because it's UI-drag-flavored
/// math, not a domain model; testable under plain `dotnet test`.
///
/// The Mac version calls `NSHapticFeedbackManager` on every successful snap (trackpad click
/// feedback) — there is no Windows equivalent and it's a side effect, not part of the mutation, so
/// it's dropped rather than ported.
public static class SnapEngine
{
    public enum SnapTargetKind
    {
        Playhead,
        ClipEdge,
        Beat,
    }

    public readonly record struct SnapTarget(int Frame, SnapTargetKind Kind);

    public readonly record struct SnapResult(int Frame, int ProbeOffset, double X);

    /// Sticky-snap multiplier: hold threshold = base threshold * this, once snapped.
    private const double StickyMultiplier = 1.5;

    /// Playhead targets get a wider catch threshold than clip edges.
    private const double PlayheadMultiplier = 1.5;

    /// Mutable state that persists across drag events for sticky snap behavior.
    public struct SnapState
    {
        public int? CurrentlySnappedTo;
        public int CurrentProbeOffset;
    }

    /// Collects all clip edges, and optionally the playhead, as snap targets. Pass
    /// `excludeClipIds` to skip clips being dragged. Pass `includePlayhead: true` when the
    /// playhead itself is NOT what's being moved.
    public static List<SnapTarget> CollectTargets(
        IReadOnlyList<Track> tracks,
        int playheadFrame = 0,
        ISet<string>? excludeClipIds = null,
        bool includePlayhead = false,
        Func<Clip, IEnumerable<int>>? beatFrames = null,
        bool includeExcludedClipBeats = false)
    {
        excludeClipIds ??= new HashSet<string>();
        var targets = new List<SnapTarget>();
        if (includePlayhead)
        {
            targets.Add(new SnapTarget(playheadFrame, SnapTargetKind.Playhead));
        }
        foreach (var track in tracks)
        {
            foreach (var clip in track.Clips)
            {
                var excluded = excludeClipIds.Contains(clip.Id);
                if (!excluded)
                {
                    targets.Add(new SnapTarget(clip.StartFrame, SnapTargetKind.ClipEdge));
                    targets.Add(new SnapTarget(clip.EndFrame, SnapTargetKind.ClipEdge));
                }
                if (beatFrames is not null && (!excluded || includeExcludedClipBeats))
                {
                    foreach (var frame in beatFrames(clip))
                    {
                        targets.Add(new SnapTarget(frame, SnapTargetKind.Beat));
                    }
                }
            }
        }
        return targets;
    }

    /// Snap position(s) to nearest target, with sticky behavior and playhead priority. Tests one
    /// or more probe positions (e.g., clip start and end) against all targets.
    public static SnapResult? FindSnap(
        int position,
        IReadOnlyList<SnapTarget> targets,
        ref SnapState state,
        double baseThreshold,
        double pixelsPerFrame,
        IReadOnlyList<int>? probeOffsets = null)
    {
        probeOffsets ??= [0];
        var baseFrameThreshold = baseThreshold / pixelsPerFrame;

        // Sticky: stay snapped until moved 2.5x threshold away.
        if (state.CurrentlySnappedTo is { } snapped)
        {
            var holdThreshold = baseFrameThreshold * StickyMultiplier;
            var probePos = position + state.CurrentProbeOffset;
            if (Math.Abs((double)(probePos - snapped)) <= holdThreshold && targets.Any(t => t.Frame == snapped))
            {
                return new SnapResult(snapped, state.CurrentProbeOffset, snapped * pixelsPerFrame);
            }
            state.CurrentlySnappedTo = null;
            state.CurrentProbeOffset = 0;
        }

        // Find closest (probe, target) pair.
        (int ProbeOffset, SnapTarget Target, double Distance)? best = null;
        foreach (var probeOffset in probeOffsets)
        {
            var probePos = position + probeOffset;
            foreach (var target in targets)
            {
                var threshold = target.Kind switch
                {
                    SnapTargetKind.Playhead => baseFrameThreshold * PlayheadMultiplier,
                    _ => baseFrameThreshold,
                };
                var dist = Math.Abs((double)(probePos - target.Frame));
                if (dist <= threshold && dist < (best?.Distance ?? double.PositiveInfinity))
                {
                    best = (probeOffset, target, dist);
                }
            }
        }

        if (best is not { } b)
        {
            return null;
        }
        state.CurrentlySnappedTo = b.Target.Frame;
        state.CurrentProbeOffset = b.ProbeOffset;
        return new SnapResult(b.Target.Frame, b.ProbeOffset, b.Target.Frame * pixelsPerFrame);
    }
}
